docs/docs/use_cases/rag/code_understanding.mdx

```diff
@@ -0,0 +1,386 @@
+import { BaseMessage } from "langchain/schema";
+import { StringOutputParser } from "langchain/schema/output_parser";
+```
+
+First we construct the combine documents chain. This chain takes in the question, chat history (from memory) and context (relevant documents). It then passes all three into the prompt which answers the users question.
+
+```typescript
+const combineDocumentsPrompt = ChatPromptTemplate.fromMessages([
+  [
+    "ai",
+    `Use the following pieces of context to answer the question at the end. If you don't know the answer, just say that you don't know, don't try to make up an answer.\n\n{context}\n\nQuestion: {question}\n`,
+  ],
+  new MessagesPlaceholder("chat_history"),
+  ["ai", "Helpful answer:"],
+]);
+```
+
+```typescript
+const combineDocumentsChain = RunnableSequence.from([
+  {
+    question: (i: { question: string }) => i.question,
+    chat_history: async () => {
+      const { chat_history } = await memory.loadMemoryVariables({});
+      return chat_history;
+    },
+    context: async (i: { question: string }) => {
+      const relevantDocs = await retriever.getRelevantDocuments(i.question);
+      return formatDocumentsAsString(relevantDocs);
+    },
+  },
+  combineDocumentsPrompt,
+  model,
+  new StringOutputParser(),
+]);
+```
+
+Then, we define our question generator chain. This is used to regenerate the users question based on the chat history. Using LLMs to regenerate questions has been shown to lead to better outputs by LLMs since it can be updated to include past context.
+
+```typescript
+const question_generator_template = ChatPromptTemplate.fromMessages([
+  [
+    "ai",
+    `Given the following conversation about a codebase and a follow up question, rephrase the follow up question to be a standalone question.`,
+  ],
+  new MessagesPlaceholder("chat_history"),
+  [
+    "ai",
+    `Follow Up Input: {question}
+Standalone question:`,
+  ],
+]);
+```
+
+```typescript
+const questionGeneratorChain = RunnableSequence.from([
+  {
+    question: (i: { question: string }) => i.question,
+    chat_history: async () => {
+      const { chat_history } = await memory.loadMemoryVariables({});
+      return chat_history;
+    },
+  },
+  question_generator_template,
+  model,
+  new StringOutputParser(),
+]);
+```
+
+In order to use the chain we defined above we create a new chain: `questionGeneratorAndCombineDocumentsChain`.
+
+This chain is called whenever there is chat history so it can first rephrase the question, then pass it through to `combineDocumentsChain` for an answer.
+
+```typescript
+const questionGeneratorAndCombineDocumentsChain = RunnableSequence.from([
+  {
+    question: (i: { question: string }) => {
+      return questionGeneratorChain.invoke(i);
+    },
+  },
+  combineDocumentsChain,
+]);
+```
+
+Finally, we'll patch it all together using a [`RunnableBranch`](/docs/expression_language/how_to/routing#using-a-runnablebranch).
```

Comment:
I think simpler to not include this? Or include it as an addendum at the end?

Reply:
We don't necessarily **need** it, but in order to be able to ask follow ups we need the branch to check if there is chat history or not. Although we could omit it, and have the followup explicitly call `questionGeneratorAndCombineDocumentsChain`, it's a bad practice compared to using a branch and I don't think we should use it in our use cases.

Reply:
Just always have the rephrase step in the initial version?

Reply:
Ah yeah that would work. Refactoring rn

Reply:
Pushed